evm/p2p/discovery.py

```diff
@@ -0,0 +1,320 @@
+        self.kademlia.recv_find_node(node, target)
+
+    def send_ping(self, node):
+        self.logger.debug('>>> pinging {}'.format(node))
+        version = rlp.sedes.big_endian_int.serialize(PROTO_VERSION)
+        payload = [version, self.address.to_endpoint(), node.address.to_endpoint()]
+        message = _pack(CMD_PING.id, payload, self.privkey)
+        self.send(node, message)
+        return message[:32]  # return the MDC to identify pongs
+
+    def send_find_node(self, node, target_node_id):
+        target_node_id = int_to_big_endian(
+            target_node_id).rjust(kademlia.k_pubkey_size // 8, b'\0')
+        self.logger.debug('>>> find_node to {}'.format(node))
+        message = _pack(CMD_FIND_NODE.id, [target_node_id], self.privkey)
+        self.send(node, message)
+
+    def send_pong(self, node, token):
+        self.logger.debug('>>> ponging {}'.format(node))
+        payload = [node.address.to_endpoint(), token]
+        message = _pack(CMD_PONG.id, payload, self.privkey)
+        self.send(node, message)
+
+    def send_neighbours(self, node, neighbours):
+        nodes = []
+        neighbours = sorted(neighbours)
+        for n in neighbours:
+            l = n.address.to_endpoint() + [n.pubkey]
+            nodes.append(l)
+
+        max_neighbours = self._get_max_neighbours_per_packet()
+        if len(nodes) <= max_neighbours:
+            message = _pack(CMD_NEIGHBOURS.id, [nodes], self.privkey)
+            self.logger.debug('>>> neighbours to {}: {}'.format(node, neighbours))
+            self.send(node, message)
+        else:
+            for i in range(0, len(nodes), max_neighbours):
+                message = _pack(CMD_NEIGHBOURS.id, [nodes[i:i + max_neighbours]], self.privkey)
+                self.logger.debug('>>> neighbours to {}: {}'.format(
+                    node, neighbours[i:i + max_neighbours]))
+                self.send(node, message)
+
+
+@to_list
+def _extract_nodes_from_payload(payload):
+    for item in payload:
+        ip, udp_port, tcp_port, node_id = item
+        address = kademlia.Address.from_endpoint(ip, udp_port, tcp_port)
+        yield kademlia.Node(node_id, address)
+
+
+def _pack(cmd_id, payload, privkey):
+    """Create and sign a UDP message to be sent to a remote node.
+
+    See https://github.com/ethereum/devp2p/blob/master/rlpx.md#node-discovery for information on
+    how UDP packets are structured.
+    """
+    cmd_id = force_bytes(chr(cmd_id))
+    expiration = rlp.sedes.big_endian_int.serialize(int(time.time() + EXPIRATION))
+    encoded_data = cmd_id + rlp.encode(payload + [expiration])
+    signature = get_ecc_backend().ecdsa_sign(encoded_data, privkey)
+    mdc = keccak(signature + encoded_data)
+    return mdc + signature + encoded_data
+
+
+def _unpack(message):
+    """Unpack a UDP message received from a remote node.
+
+    Returns the public key used to sign the message, the cmd ID, payload and mdc.
+    """
+    mdc = message[:MAC_SIZE]
```

Comment:
What does `mcd` stand for?

Reply:
I couldn't find, but it comes from https://github.com/ethereum/go-ethereum/wiki/RLPx-----Node-Discovery-Protocol. It's the hash of the whole message, so I'll rename it to hash as that's certainly more meaningful than mdc

Reply:
*naming nitpick*  

`message_hash` rather than the slightly awkward avoidance of the built in `hash` function via `hash_`?

Reply:
Sure, changed